1-Draft/RFCNNNN-String-Manipulating-Cmdlets.md

```diff
@@ -0,0 +1,84 @@
+---
+RFC: RFCNNNN
+Author: Friedrich Weinmann
+Status: Draft
+SupercededBy: n/a
+Version: 1.0
+Area: cmdlets/data manipulation
+Comments Due: 
+Plan to implement: Yes
+---
+
+# String Manipulating Cmdlets
+
+Add cmdlets that manipulate strings and are capable of doing so on the pipeline.
+
+## Motivation
+
+Existing tools to manipulate strings - the most common data type manipulated in PowerShell are lacking in usability.
+The .NET tools available through string and regex are inconvenient to use.
+String-manipulating operators (such as `-split`) are unintuitive to explore, as their options are not easily discoverable.
+Neither option allows use on the pipeline, forcing interruption of the pipeline or use of inefficient `ForEach-Object` calls.
+
+From this come two main pain points this RFC is aimed to resolve:
+
+Current string data manipulation ...
+
+ - is technically insufficient from a performance perspective.
+ - is unintuitive and inconvenient for the end user.
+
```

Comment:
Now that we have an rapidly increasing number of new users to PowerShell, discover-ability is a significantly more important issue.

Operators have always been difficult to discover, and have had limited documentation.

Personally I find ```-f``` to be extremely un-PowerShell. It doesn't clearly state what it does or how it works.

Reply:
As to the statement "I haven't seen any modules that do this". We don't always release all of our modules to the public. There are probably dozens of people maintaining similar modules with CMDLets similar to that which has been described.

Reply:
I think there are multiple groups of users who utilize PowerShell.  For the "Programmers", I am sure that a plethora of cmdlets is a bad thing.  Also they probably well know all of the quirks of every PowerShell command.  However, I would strongly suggest there is a group of users who may be termed "Casual" or "System Admins" for whom a plethora of cmdlets is great.  They mainly write one or two line scripts that are just pumping data through a pipeline.  For these users, the cmdlets described here are great.  Are there other ways to get there, sure, but as PowerShell branches out to serve more OSs and more user groups, I think ease of use and discoverability is key.  My 2cents.

Reply:
@iSazonov - just because there's no PowerShell module that you can find publicly doesn't mean that:

(1) Folks don't write their own janky implementations internally
(2) Folks have time to write something that presumably might better be handled by the PowerShell team
(3) Folks don't need it or wouldn't use it

It's always painful using some of the workarounds in code - it looks ugly af, and it's harder to follow for many folks

Reply:
It is **Simply NOT TRUE** that nobody has created a "popular module" with similar cmdlets.  PSCX was the original super-popular module. It was the one we told **everyone** to get, and many people are still using it as a matter of course. It still has compiled `Join-String` and `Split-String` commands (and `Edit-File` for regex replace, as well as related commands like `ConvertTo-Base64` and `Get-Clipboard` etc.).

In fact, @FriedrichWeinmann, if PowerShell core is going to stomp on **yet another** set of PSCX commands, I would like to suggest that *this time* you should make sure your implementation _matches_ what's in PSCX (or be additive to it) instead of creating another source of frustration like `Expand-Archive`.

Reply:
It's also worth pointing out that setting aside PSCX there are **other** implementations of `Join-String` and `Split-String` and `Update-String` (actually `Update-StringInFile` and `Edit-File`) and `Format-String` (no less than [four of these](https://www.powershellgallery.com/items?q=Format-String), 2 by MVPs and 2 by MSFT) on the PowerShell gallery...

Frankly though, I'm not sure that argues _for_ including this in PowerShell, rather than shipping it externally.

Reply:
@iSazonov 
> If we talk about performance, .Net methods will be faster, and we can even use C# and assemblies by means of Add-Type.

Following your logic, if C# is better, why should PowerShell even exist? And the answer is the vast majority of PowerShell users _aren't_ programmers. Telling them to use `Add-Type` is completely unreasonable. In terms of _relative_ performance, this
```powershell
@('"a","b","c"') * 1000 | Split-String '","' -trim '"'
```
is both easier to use and faster than the alternative:
```powershell
@('"a","b","c"') * 1000 | foreach { $_.Trim('"') }
```
> Now I have thousands of cmdlets on my computer and it is no longer working.

And .NET has tens of thousands of properties/methods, etc. Is that not working for you either? How about the PowerShell repo? It has somewhere around 6000 files (including build artifacts). Too many? Anyway, I agree that we need a better way to search for relevant commands but that is orthogonal to this discussion.  (but I would certainly be happy to look at an RFC proposing ways to improve the current state of affairs.)
> We have a powerful format system. Why should we have Format-String?

Per the  RFC text, this cmdlet has nothing to do with Output & Formatting. It's a wrapper around `String.Format()` (or `-f`) but works in the pipeline on a stream of objects. Right now you have to do:
```powershell
gci | ForEach-Object { 'name: {0} length' -f $_.name, $_.length }
```
With `Format-String` you just do:
```powershell
gci | Format-String  -FormatString 'name: {0} length' -property name,length
```
This is easier, more discoverable and certainly faster as there is no `ScriptBlock` dispatch.